Adding QOS in response to LinkADRReq and fixing class C bugs

hasnainvirk opened the pull request:
### Description

Adds QOS mechanism for unconfirmed messages as per 1.0.2 spec in response to LinkADRReq mac command.  In addition to that it fixes various bugs potentially hindering class C operation. 

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [X] Fix
    [] Refactor
    [ ] Target update
    [] Functionality change
    [ ] Breaking change

Commit:
Style correction

LinkADRReq parameters and certain parameters used camel case
which is not the recommended style.

Commit:
Adding check for ongoing automatic-uplink

Before going after an automatic uplink, we should check if there was an
automatic uplink already ongoing, i.e., the ack for the previous
automatic uplink cycle has not been received.
If there is we shouldn't queue the new automatic uplink and wait for
the previous Ack cycle to complete.

Commit:
Adding QOS handling and fixing bugs for Class C

LinkADRReq mac command can be used by the network server to set a
certain level of QOS using NbTrans field which is applicable to
Unconfirmed traffic only for 1.0.2 spec.
This commit introduces mechanisms to facilitate this QOS. It means to
repeat an outgoing unconfirmed message NbTrans times without changing
its frame counter.

For class C, we have retired the ack_expiry_timer_for_class_c and have
replaced it with another timer which mimics the RX2 closure as in Class
A but doesn't actually close RX2 window. It's just a mechanism by which
the state machine is informed that the you can proceed forward, we have
not received anything in RX2 window either. This is needed as RX2
doesn't timeout in class C (i.e., the radio remains in continuous mode).
In addition to that we need to close any pending timers for Receive
windows after the MIC has passed and the Duplicate counter check has
also been passed.

hasnainvirk commented:
@kivaisan @AnttiKauppila @kjbracey-arm Please review.

hasnainvirk commented:
Tests added here 
https://github.com/ARMmbed/mbed-lora-tests-simulator-private/pull/6 [Simulator test]
https://github.com/ARMmbed/mbed-lora-tests-private/pull/59 [Functional HW test]

kivaisan commented on features/lorawan/LoRaWANStack.cpp:
```diff
@@ -600,39 +601,110 @@ void LoRaWANStack::process_transmission(void)
         if (_loramac.get_mcps_confirmation()->req_type == MCPS_CONFIRMED) {
-            _ctrl_flags |= TX_ONGOING_FLAG;
-            _ctrl_flags &= ~TX_DONE_FLAG;
+            //_ctrl_flags &= ~TX_DONE_FLAG;
```
Remove dead code

kivaisan commented on features/lorawan/LoRaWANStack.cpp:
```diff
@@ -600,39 +601,110 @@ void LoRaWANStack::process_transmission(void)
-            // application when RX2 windows is elapsed, i.e., in process_reception_timeout()
-            _ctrl_flags &= ~TX_ONGOING_FLAG;
+        } else {
+            //_ctrl_flags |= TX_DONE_FLAG;
```
Dead code here too

kivaisan reviewed (approved):
LGTM. Just few cosmetics.

hasnainvirk commented:
Dead code removed, rebased.

Commit:
Making sure that RX slots open after state change

After transmission we should change the state before invoking opening of
slots as we may start receiving in the rx slots and the state would
suddenly change from SENDING to RECEIVING without going through the
ACK_WAIT state (in case of CONFIRMED messages). Tests show that after
this slight adjustment, our number of ack retries have significantly
reduced.

Commit:
TX post-process for CONFIRMED UL in no-reception case

The idea behind the method post_process_no_reception() is to post
process any outgoing TX but we shouldn't do that if a CONFIRMED message
is outgoing and there are still some retries left.

hasnainvirk commented:
@kivaisan @AnttiKauppila @kjbracey-arm Please review again. Added a few commits for stability improvement and rebased after few PRs went in.

kivaisan commented on features/lorawan/lorastack/mac/LoRaMac.cpp:
```diff
@@ -889,6 +889,8 @@ void LoRaMac::on_backoff_timer_expiry(void)
 
 void LoRaMac::open_rx1_window(void)
 {
+    tr_debug("Opening RX1 Window");
```
No. Opening RX windows is quite time critical so let's not print until the window has been opened.

kivaisan commented on features/lorawan/lorastack/mac/LoRaMac.cpp:
```diff
@@ -909,12 +911,12 @@ void LoRaMac::open_rx1_window(void)
 
 void LoRaMac::open_rx2_window()
 {
+    tr_debug("Opening RX2 Window, Frequency = %lu", _params.rx_window2_config.frequency);
```
The same issue here.

Commit:
Proper handling of RX1 frequency in rx_config

Previously, we weren't filling in RX1 frequecny in rx_window1_config
structure. However, everything worked as in LoRaPHY::rx_config() API
there was a check which filled in correct RX1 frequency.
Now we are filling in RX1 freq. properly while we are computing
parameters for RX1 window.

Commit:
Fixing coverity findings

A couple of the coverity analysis findings are being treated here. For
the rest there will be a separate PR.

hasnainvirk commented:
@kivaisan @kjbracey-arm @AnttiKauppila @0xc0170 Please review again. Tests updated as well in their respective PRs. Some corner cases are covered and improvements in stability.

AnttiKauppila commented:
@0xc0170 Canthis be built?

NirSonnenschein commented:
/morph build

0xc0170 commented:
/morph mbed2-build
